packages/truffle-migrate/migration.js

```diff
@@ -0,0 +1,188 @@
+const Deployer = require("truffle-deployer");
+const Require = require("truffle-require");
+const Emittery = require('emittery');
+const async = require("async");
+const Web3 = require("web3");
+
+const ResolverIntercept = require("./resolverintercept");
+
+const util = require('util');
+
+class Migration {
+
+  constructor(file, reporter, options){
+    this.file = path.resolve(file);
+    this.reporter = reporter;
+    this.number = parseInt(path.basename(file));
+    this.emitter = new Emittery();
+    this.isFirst = false;
+    this.isLast = false;
+    this.dryRun = options.dryRun;
+    this.interactive = options.interactive;
+    this.options = options || {};
+  }
+
+  // ------------------------------------- Private -------------------------------------------------
+  /**
+   * Loads & validates migration, then runs it.
+   * @param  {Object}   options  config and command-line
+   * @param  {Object}   context  web3
+   * @param  {Object}   deployer truffle module
+   * @param  {Object}   resolver truffle module
+   * @param  {Function} callback
+   */
+  async _load(options, context, deployer, resolver, callback){
+    const self = this;
+
+    // Load assets and run `execute`
+    try {
+      const accounts = await context.web3.eth.getAccounts();
+      const requireOptions = {
+        file: self.file,
+        context: context,
+        resolver: resolver,
+        args: [deployer],
+      }
+
+      Require.file(requireOptions, async (err, fn) => {
+        if (err) return callback(err);
+
+        const unRunnable = !fn || !fn.length || fn.length == 0;
+
+        if (unRunnable){
+          const msg = `Migration ${self.file} invalid or does not take any parameters`;
+          return callback(new Error(msg));
+        }
+
+        // `migrateFn` might be sync or async. We negotiate that difference in
+        // `execute` through the deployer API.
+        const migrateFn = fn(deployer, options.network, accounts);
+        await self._deploy(options, deployer, resolver, migrateFn, callback);
+      });
+
+    } catch(err){
+      callback(err)
+    }
+  }
+
+  /**
+   * Initiates deployer sequence, then manages migrations info
+   * publication to chain / artifact saving.
+   * @param  {Object}   options     config and command-line
+   * @param  {Object}   deployer    truffle module
+   * @param  {Object}   resolver    truffle module
+   * @param  {[type]}   migrateFn   module.exports of a migrations.js
+   */
+  async _deploy(options, deployer, resolver, migrateFn, callback) {
+    const self = this;
+
+    try {
+      await deployer.start();
+
+      // Allow migrations method to be async and
+      // deploy to use await
+      if (migrateFn && migrateFn.then !== undefined){
+        await deployer.then(() => migrateFn);
+      }
+
+      // Migrate without saving
+      if (options.save === false) return;
+
+      // Write migrations record to chain
+      const Migrations = resolver.require("./Migrations.sol");
```

Comment:
`resolver.require("Migrations")`

Reply:
👍

Reply:
Hmmm. Need to investigate this a bit. Is the hard-coding of the path (carried over from original) the cause of the bug identified [here](https://github.com/trufflesuite/truffle/issues/713#issuecomment-403031560). Or.....?

Reply:
It's just more canonical. Artifacts are by `contract` name, not filename.

Reply:
The older migrate tests at `truffle-core` crash when this change is made. I want to deal with separately from this PR as a bug in the resolver - not sure what's going here....